[clang-tidy] Add bugprone-move-shared-pointer-contents check.

pizzud opened the pull request:
This check detects moves of the contents of a shared pointer rather than the pointer itself. Other code with a reference to the shared pointer is probably not expecting the move.

The set of shared pointer classes is configurable via options to allow individual projects to cover additional types.

pizzud committed:
[clang-tidy] Add bugprone-move-shared-pointer-contents check.

This check detects moves of the contents of a shared pointer rather than the
pointer itself. Other code with a reference to the shared pointer is probably
not expecting the move.

The set of shared pointer classes is configurable via options to allow individual
projects to cover additional types.

pizzud committed:
[clang-tidy] Add bugprone-move-shared-pointer-contents check.

This check detects moves of the contents of a shared pointer rather than the
pointer itself. Other code with a reference to the shared pointer is probably
not expecting the move.

The set of shared pointer classes is configurable via options to allow individual
projects to cover additional types.

pizzud committed:
[clang-tidy] Add bugprone-move-shared-pointer-contents check.

This check detects moves of the contents of a shared pointer rather than the
pointer itself. Other code with a reference to the shared pointer is probably
not expecting the move.

The set of shared pointer classes is configurable via options to allow individual
projects to cover additional types.

pizzud committed:
[clang-tidy] Add bugprone-move-shared-pointer-contents check.

This check detects moves of the contents of a shared pointer rather than the
pointer itself. Other code with a reference to the shared pointer is probably
not expecting the move.

The set of shared pointer classes is configurable via options to allow individual
projects to cover additional types.

5chmidti reviewed (commented):
I noticed a problem with your matcher, so I reviewed the rest of it while I was at it.

The problem is that you do not consider a type-dependent `std::shared_ptr<T>` and the following test case fails:

```c++
template <typename T>
void dependentType() {
  std::shared_ptr<T> p;
  T y = std::move(*p);
  // CHECK-FIXES: *std::move(p)
}
// CHECK-MESSAGES: :[[@LINE-3]]:11: warning: don't move the contents out of a shared pointer, as other accessors expect them to remain in a determinate state [bugprone-move-shared-pointer-contents]
```

In this case, your `callee` is no longer a `functionDecl` but an `unresolvedLookupExpr` and the `*` operator is a `UnaryOperator` instead of a `cxxOperatorCallExpr`.

Otherwise, looks like a good check to have.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.h:
```diff
@@ -0,0 +1,37 @@
+#define LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_BUGPRONE_MOVESHAREDPOINTERCONTENTSCHECK_H
+
+#include <vector>
+
```
Excessive newline.

pizzud replied:
Done. Along with some unexpected shuffles by clang-format.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.h:
```diff
@@ -0,0 +1,37 @@
+  isLanguageVersionSupported(const LangOptions &LangOptions) const override;
+
+private:
+  const std::vector<StringRef> SharedPointerClasses;
```
Please include `llvm/StringRef.h`.

pizzud replied:
Done.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/move-shared-pointer-contents.rst:
```diff
@@ -0,0 +1,17 @@
+.. option :: SharedPointerClasses
+
+   A semicolon-separated list of class names that should be treated as shared
+   pointers. By default only `std::shared_ptr` is included.
```
I think , `Default is `std::shared_ptr`.` would be better.

pizzud replied:
Done.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,60 @@
+  const auto *Call = Result.Nodes.getNodeAs<CallExpr>("call");
+  const auto *Op = Result.Nodes.getNodeAs<Expr>("op");
+
+  if (Op) {
```
Done.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.h:
```diff
@@ -0,0 +1,37 @@
+
+namespace clang::tidy::bugprone {
+
+/// FIXME: Write a short description.
```
Done.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,60 @@
+      SharedPointerClasses(utils::options::parseStringList(
+          Options.get("SharedPointerClasses", "std::shared_ptr"))) {}
+
+MoveSharedPointerContentsCheck::~MoveSharedPointerContentsCheck() = default;
```
Done.

pizzud commented:
> I noticed a problem with your matcher, so I reviewed the rest of it while I was at it.
> 
> The problem is that you do not consider a type-dependent `std::shared_ptr<T>` and the following test case fails:
> 
> ```c++
> template <typename T>
> void dependentType() {
>   std::shared_ptr<T> p;
>   T y = std::move(*p);
>   // CHECK-FIXES: *std::move(p)
> }
> // CHECK-MESSAGES: :[[@LINE-3]]:11: warning: don't move the contents out of a shared pointer, as other accessors expect them to remain in a determinate state [bugprone-move-shared-pointer-contents]
> ```
> 
> In this case, your `callee` is no longer a `functionDecl` but an `unresolvedLookupExpr` and the `*` operator is a `UnaryOperator` instead of a `cxxOperatorCallExpr`.
> 
> Otherwise, looks like a good check to have.

I'm really struggling with this, and also with getting the git history right after three months oops. The UnaryOperator is matchable without much difficulty, but there appears to be no way to introspect the UnresolvedLookupExpr to ensure I'm matching calls to `std::move`.

5chmidti commented:
> but there appears to be no way to introspect the UnresolvedLookupExpr to ensure I'm matching calls to `std::move`

Try searching for `unresolvedLookupExpr` & `UnresolvedLookupExpr` in the clang-tidy directory. There is a high probability that another check has done something related/similar.

https://github.com/llvm/llvm-project/blob/2b7191c8993b5608ddb8b89c049717b497265796/clang-tools-extra/clang-tidy/cppcoreguidelines/RvalueReferenceParamNotMovedCheck.cpp#L46-L47

And there is `OverloadExpr::decls`, which `UnresolvedLookupExpr` inherits (https://clang.llvm.org/doxygen/classclang_1_1UnresolvedLookupExpr.html).

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,51 @@
+void MoveSharedPointerContentsCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      callExpr(
+          callee(functionDecl(hasName("std::move"))),
```
consider hasName("::std::move")

pizzud replied:
Done.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,51 @@
+    StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      SharedPointerClasses(utils::options::parseStringList(
+          Options.get("SharedPointerClasses", "std::shared_ptr"))) {}
```
Same here, consider ::std::shared_ptr, you could add also ::boost::shared_ptr

pizzud replied:
Switched to ::. Is the Boost version really still so widely used to justify this in 2023 when std::shared_ptr has been around since C++11?

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.h:
```diff
@@ -0,0 +1,39 @@
+  isLanguageVersionSupported(const LangOptions &LangOptions) const override {
+    return LangOptions.CPlusPlus11;
+  }
+
```
Consider adding here:
```
std::optional<TraversalKind> getCheckTraversalKind() const override {
 return TK_IgnoreUnlessSpelledInSource;
}

```

to exclude implicit code., or add explicitly with TK_AsIs to avoid changes in future, differences would be in template instances.

pizzud replied:
I've been warned off IgnoreUnlessSpelledInSource in the past since it won't match hits in headers, only .cc files, and we now want to catch template instantiations. I'll use TK_AsIs.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,51 @@
+  Finder->addMatcher(
+      callExpr(
+          callee(functionDecl(hasName("std::move"))),
+          hasArgument(0, cxxOperatorCallExpr(hasOverloadedOperatorName("*"),
```
Consider adding support for "*(x.get())", but overall it's fine.

pizzud replied:
The current implementation actually catches this by dint of the unaryOperator(hasOperatorName("*")). I have some fear it'll be overbroad but haven't managed to construct a problematic case yet.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/CMakeLists.txt:
```diff
@@ -25,6 +25,7 @@ add_clang_library(clangTidyBugproneModule
   ImplicitWideningOfMultiplicationResultCheck.cpp
   InaccurateEraseCheck.cpp
   IncorrectEnableIfCheck.cpp
+  MoveSharedPointerContentsCheck.cpp
```
Duplicated file, remove this line

pizzud replied:
Done, bad merge.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,51 @@
+  diag(Call->getBeginLoc(),
+       "don't move the contents out of a shared pointer, as other accessors "
+       "expect them to remain in a determinate state")
+      << FixItHint::CreateInsertion(Call->getBeginLoc(), "*")
```
Remove fix, this isn't a valid one. Example:

```
#include <memory>

struct A{};

void test(A&&);

void w()
{
    std::shared_ptr<A> a;
    test(*std::move(a));
} 
```

this won't compile, operator * will return lvalue reference anyway.

pizzud replied:
Done.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/list.rst:
```diff
@@ -118,6 +119,7 @@ Clang-Tidy Checks
    :doc:`bugprone-posix-return <bugprone/posix-return>`, "Yes"
    :doc:`bugprone-redundant-branch-condition <bugprone/redundant-branch-condition>`, "Yes"
    :doc:`bugprone-reserved-identifier <bugprone/reserved-identifier>`, "Yes"
+   :doc:`bugprone-shared-pointer-contents-move <bugprone/shared-pointer-contents-move>`, "Yes"
```
reuplicated check, remove

pizzud replied:
Done.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/list.rst:
```diff
@@ -107,6 +107,7 @@ Clang-Tidy Checks
    :doc:`bugprone-misplaced-pointer-arithmetic-in-alloc <bugprone/misplaced-pointer-arithmetic-in-alloc>`, "Yes"
    :doc:`bugprone-misplaced-widening-cast <bugprone/misplaced-widening-cast>`,
    :doc:`bugprone-move-forwarding-reference <bugprone/move-forwarding-reference>`, "Yes"
+   :doc:`bugprone-move-shared-pointer-contents <bugprone/move-shared-pointer-contents>`, "Yes"
```
remove "Yes" and do not provide fixes for this check.

pizzud replied:
Done.

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/move-shared-pointer-contents.cpp:
```diff
@@ -0,0 +1,75 @@
+
+void configWorks() {
+  my::OtherSharedPtr<int> p;
+  int x = std::move(*p);
```
Use some nontrivial type instead of int. std::move on int is pointless anyway.

pizzud replied:
Done, though my full rule-of-five implementation might be overkill.

PiotrZSL reviewed (changes requested):
Few fixes needed, but looks you are on good road.

pizzud commented:
Try searching for `unresolvedLookupExpr` & `UnresolvedLookupExpr` in the clang-tidy directory. There is a high probability that another check has done something related/similar.
> 
> https://github.com/llvm/llvm-project/blob/2b7191c8993b5608ddb8b89c049717b497265796/clang-tools-extra/clang-tidy/cppcoreguidelines/RvalueReferenceParamNotMovedCheck.cpp#L46-L47

Thanks, this was super helpful and exactly what I was looking for!

I think the missing insight was that I could go from the lookup -> decls.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,49 @@
+                                        callee(cxxMethodDecl(ofClass(
+                                            matchers::matchesAnyListedName(
+                                                SharedPointerClasses))))),
+                                    unaryOperator(hasOperatorName("*")))))
```
What you need is something like:
```
callExpr(callee(unresolvedLookupExpr()), hasArgument(0, unaryOperator(hasUnaryOperand(expr(hasType(namedDecl().bind("c")))))))
```

But additionally:
- you need to ignore some implicit code in here, mainly in hasArgument(0, ...)
- you need take into account that type may be typedef to shared_ptr (you need to get canonical type)
- you need take into account that type may be a reference to shared_ptr (you may need to drop reference)

As for get, it's actually easy to handle:
```
CallExpr 0x55e2d6219a18 <col:10, col:28> '<dependent type>'
            |-UnresolvedLookupExpr 0x55e2d6219928 <col:10, col:15> '<overloaded function type>' lvalue (no ADL) = 'move' 0x55e2d5a7c948 0x55e2d5c63b08
            `-UnaryOperator 0x55e2d6219a00 <col:20, col:27> '<dependent type>' prefix '*' cannot overflow
              `-CallExpr 0x55e2d62199e0 <col:21, col:27> '<dependent type>'
                `-CXXDependentScopeMemberExpr 0x55e2d6219998 <col:21, col:23> '<dependent type>' lvalue .get
                  `-DeclRefExpr 0x55e2d6219978 <col:21> 'std::shared_ptr<T>':'shared_ptr<T>' lvalue ParmVar 0x55e2d6219608 'p' 'std::shared_ptr<T>':'shared_ptr<T>'
```

pizzud replied:
Took me a bit to get a handle on it, but I think I have these cases handled. Thanks for the pointers (pun very much intended!).

PiotrZSL reviewed (commented):
Few issues with `unaryOperator` left, after that will be fixed, it would look fine.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+  // typedefs.
+
+  bool found = false;
+  auto *Template = llvm::dyn_cast<TemplateSpecializationType>(
```
chould be `const auto *`.

pizzud replied:
Done, assuming all my pushes went through as intended.

PiotrZSL replied:
They didnt...

pizzud replied:
Okay, this time it made it.

pizzud commented:
Not sure how to squash properly; I tried rebasing but all the commits on the main branch snuck in to the history and I'm quite new to git. If it would be preferred I can store this off as a patch and open up a new PR. Sorry for all the hassle with the merges!

EugeneZelenko commented:
You need to do interactive rebase:

```
git fetch --prune # get latest main and remove local copies deleted branch 
git rebase origin/main # rebase from main
git rebase -i origin/main # interactive rebase, just use fixup command
git rebase --force origin <your branch name> # push updated branch
```

pizzud commented:
A coworker was able to help me get this working by rebasing against upstream, so I think we're good to go at this point. Thanks for all the feedback!

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+  clang::SourceLocation Loc;
+  if (Unresolved) {
+    Loc = UnresolvedCall->getBeginLoc();
+  } else if (UnresolvedGet) {
```
there should be here also check for UnresolvedGetPtr

pizzud replied:
Done. They "should" only appear together, but that's a load-bearing "should" if ever I've seen one. :)

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+  const bool UnresolvedGet = isSharedPointerClass(UnresolvedGetPtr);
+
+  clang::SourceLocation Loc;
+  if (Unresolved) {
```
there should be here also check for UnresolvedGetCall

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+    Loc = GetCall->getBeginLoc();
+  } else if (Call != nullptr) {
+    Loc = Call->getBeginLoc();
+  }
```
else return;

pizzud replied:
Done.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+
+void MoveSharedPointerContentsCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *Call = Result.Nodes.getNodeAs<CallExpr>("call");
+  const auto *GetCall = Result.Nodes.getNodeAs<CallExpr>("get_call");
```
what the point of binding those 2 to difrent names, when usage is same, just bind both of them to "call", and then one if less, same with those unresolved, just bind them both to unresolved_call

pizzud replied:
Sorry, I don't follow this. "call" and "get_call" are bound nodes from different matchers, so they need to be checked separately, so they need to be looked up separately in `Result.Nodes`, and the same for the unresolved versions. We need both to a) figure out which case we're in and b) report at the beginning of the  I removed the separate variables in favor of using C++17 if-statement-initializers, which also looks like a small performance improvement from doing fewer unconditional lookups.

pizzud replied:
Obsolete.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+    for (const llvm::StringRef SharedPointer : SharedPointerClasses) {
+      // SharedPointer entries may or may not have leading ::, but TypeName
+      // definitely won't.
+      if (SharedPointer == TypeName || SharedPointer.substr(2) == TypeName) {
```
I don't think this works, unless I'm missing something. matchesAnyListedName requires a NamedDecl, but when the template type is unresolved we don't have a NamedDecl, just a DeclRefExpr.

PiotrZSL replied:
getAsTemplateDecl returns NamedDecl, all you need to do is just move lines 83-109 into AST_MATCHER_P with extra InnerMatcher that would be used to check name. There are some checks that do similar stuff.

pizzud replied:
Thanks for the pointer, I finally managed to figure this out.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+
+bool MoveSharedPointerContentsCheck::isSharedPointerClass(
+    const QualType *QT) const {
+  if (QT == nullptr) {
```
Done in a way that will hopefully avoid goto-fail issues down the line.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+
+bool MoveSharedPointerContentsCheck::isSharedPointerClass(
+    const VarDecl *VD) const {
+  if (VD == nullptr) {
```
Done in a way that will hopefully avoid goto-fail issues down the line.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+      hasAnyDeclaration(namedDecl(hasUnderlyingDecl(hasName("::std::move"))))));
+
+  // Unresolved type, direct move.
+  Finder->addMatcher(
+      callExpr(
+          isStdMoveUnresolved,
+          hasArgument(0, unaryOperator(hasOperatorName("*"),
+                                       hasUnaryOperand(declRefExpr(hasType(
+                                           qualType().bind("unresolved_p")))))))
+          .bind("unresolved_call"),
+      this);
+  // Annoyingly, the declRefExpr in the unresolved-move-of-get() case
+  // is of <dependent type> rather than shared_ptr<T>, so we have to
+  // just fetch the variable. This does leave a gap where a temporary
+  // shared_ptr wouldn't be caught, but moving out of a temporary
+  // shared pointer is a truly wild thing to do so it should be okay.
+
+  // Unresolved type, move out of get().
+  Finder->addMatcher(
+      callExpr(isStdMoveUnresolved,
+               hasArgument(
+                   0, unaryOperator(hasOperatorName("*"),
+                                    hasDescendant(cxxDependentScopeMemberExpr(
+                                        hasMemberName("get"))),
+                                    hasDescendant(declRefExpr(to(
+                                        varDecl().bind("unresolved_get_p")))))))
+          .bind("unresolved_get_call"),
+      this);
```
Same as above.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+  auto isStdMove = callee(functionDecl(hasName("::std::move")));
+
+  // Resolved type, direct move.
+  Finder->addMatcher(
+      callExpr(isStdMove, hasArgument(0, cxxOperatorCallExpr(
+                                             hasOverloadedOperatorName("*"),
+                                             callee(cxxMethodDecl(ofClass(
+                                                 matchers::matchesAnyListedName(
+                                                     SharedPointerClasses)))))))
+          .bind("call"),
+      this);
+
+  // Resolved type, move out of get().
+  Finder->addMatcher(
+      callExpr(
+          isStdMove,
+          hasArgument(
+              0, unaryOperator(
+                     hasOperatorName("*"),
+                     hasUnaryOperand(cxxMemberCallExpr(callee(cxxMethodDecl(
+                         hasName("get"), ofClass(matchers::matchesAnyListedName(
+                                             SharedPointerClasses)))))))))
+          .bind("get_call"),
+      this);
```
Sadly the `MatchFinder::addMatcher` overloads become ambiguous with `anyOf`. I also tried `callExpr(anyOf(allOf(...)).bind())` but that doesn't work because `anyOf` doesn't support binding.

PiotrZSL replied:
you can duplicate callExpr or expr, like `callExpr(anyOf(callExpr().bind("a"), callExpr().bind("b")))`
callExpr/expr will work like allOf

pizzud replied:
Done.

pizzud commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/move-shared-pointer-contents.rst:
```diff
@@ -0,0 +1,20 @@
+
+   A semicolon-separated list of class names that should be treated as shared
+   pointers. Classes are resolved through aliases, so any alias to the defined
+   classes will be considered. Default is `std::shared_ptr`.
```
Done.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.h:
```diff
@@ -0,0 +1,45 @@
+namespace clang::tidy::bugprone {
+
+/// Detects calls to move the contents out of a ``std::shared_ptr`` rather than
+///  moving the pointer itself.
```
Done.

pizzud commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/move-shared-pointer-contents.rst:
```diff
@@ -0,0 +1,20 @@
+
+bugprone-move-shared-pointer-contents
+=====================================
+
```
Done.

pizzud commented on clang-tools-extra/docs/clang-tidy/checks/list.rst:
```diff
@@ -118,6 +119,7 @@ Clang-Tidy Checks
    :doc:`bugprone-posix-return <bugprone/posix-return>`, "Yes"
    :doc:`bugprone-redundant-branch-condition <bugprone/redundant-branch-condition>`, "Yes"
    :doc:`bugprone-reserved-identifier <bugprone/reserved-identifier>`, "Yes"
+   :doc:`bugprone-shared-pointer-contents-move <bugprone/shared-pointer-contents-move>`, "Yes"
```
Done.

pizzud commented:
Ack, I messed this up like whoa! Sorry for all the noise and extraneous commits sneaking in.

pizzud commented:
> Ack, I messed this up like whoa! Sorry for all the noise and extraneous commits sneaking in.

Okay fixed this, but i'm not sure how to unrequest reviews.

Endilll commented:
I cleaned up the reviewers. I suggest to use `git merge` next time you need to bring your PR up to date with `main`. We are not worried about merge commits, as squash is the only available merge option for PRs.

I also suggest to use "Start review" button when you respond to multiple comments, as that reduces amount of e-mail notifications.

pizzud commented:
> I cleaned up the reviewers. I suggest to use `git merge` next time you need to bring your PR up to date with `main`. We are not worried about merge commits, as squash is the only available merge option for PRs.
> 

Thanks! I'll try merge next time.

> I also suggest to use "Start review" button when you respond to multiple comments, as that reduces amount of e-mail notifications.

Start review seems weird since the review is already in flight and I think I'd have to revert to draft in between - the button isn't currently visible. If it's less spammy I'm okay with it, though. I'm also used to a code review tool where the comments get batched into drafts and then you publish them all together when you have new code.

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,157 @@
+                   0, unaryOperator(hasOperatorName("*"),
+                                    hasDescendant(cxxDependentScopeMemberExpr(
+                                        hasMemberName("get"))),
+                                    hasDescendant(declRefExpr(to(
```
I wasn't able to manufacture one in my attempts to use shared_ptr::get as a function pointer. Is that the kind of thing you were thinking of?

PiotrZSL replied:
issue could be if you would have call expression in call expression, like:
`x = std::move(getSomething(y).field)`
Would be good to check if there are no other calls as parent.

pizzud replied:
Couldn't actually get this to fire a false-positive.

pizzud committed:
Merge remote-tracking branch 'refs/remotes/origin/shared-ptr' into shared-ptr

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/list.rst:
```diff
@@ -117,7 +117,8 @@ Clang-Tidy Checks
    :doc:`bugprone-posix-return <bugprone/posix-return>`, "Yes"
    :doc:`bugprone-redundant-branch-condition <bugprone/redundant-branch-condition>`, "Yes"
-   :doc:`bugprone-reserved-identifier <bugprone/reserved-identifier>`, "Yes"
+   :doc:`bugprone-reserved-identifier <bugprone/reserved-identifier>`, "Yes",
```
remove , at the end

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/move-shared-pointer-contents.cpp:
```diff
@@ -0,0 +1,125 @@
+}
+
+template <typename T>
+struct shared_ptr {
```
try including this header: clang-tools-extra/test/clang-tidy/checkers/modernize/Inputs/smart-ptr/shared_ptr.h instead

PiotrZSL commented:
@pizzud Just reminder that Clang-tidy 18 branch out is in ~3 days

PiotrZSL commented:
@pizzud Do you plan to finish this check, just asking as idea behind it is really good.

pizzud commented:
Yep. I've been slammed at work with other stuff and also ran into a knotty
problem I haven't managed to find a solution for: when
shared_ptr::operator* and/or get() are defined in a parent of shared_ptr,
the returned class from the matchers is the parent, not the class in the
config. This was revealed when I switched to the other check's testdata
implementation, and it's a really good thing it was because MSVC and either
libc++ or libstdc++ (I forget which) implement shared_ptr this way. I've
been meaning to ask the internal clang-tidy experts if they have any ideas,
but haven't managed to free up enough time. =/ Accessing the class with the
method definition is easy, but we have a subclass and the available methods
in CXXRecordDecl only let you go up the inheritance tree, not down. I was
hoping to get back to this later this week.

On Wed, Mar 6, 2024 at 8:42 AM Piotr Zegar ***@***.***> wrote:

> @pizzud <https://github.com/pizzud> Do you plan to finish this check,
> just asking as idea behind it is really good.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/67467#issuecomment-1981309710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5LI24JQ3UELX3YBUN57JGLYW5BQZAVCNFSM6AAAAAA5IDKJ7KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBRGMYDSNZRGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

PiotrZSL commented:
@pizzud don't worry much about this one, this can be easily solved by using hasType instead of ofClass.
Next branch out is in ~5 months, so there is some time.

pizzud committed:
This check detects moves of the contents of a shared pointer rather than the
pointer itself. Other code with a reference to the shared pointer is probably
not expecting the move.

The set of shared pointer classes is configurable via options to allow individual
projects to cover additional types.

pizzud committed:
This check detects moves of the contents of a shared pointer rather than the
pointer itself. Other code with a reference to the shared pointer is probably
not expecting the move.

The set of shared pointer classes is configurable via options to allow individual
projects to cover additional types.

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -383,7 +388,7 @@ Changes in existing checks
 
 - Improved :doc:`misc-unused-using-decls
   <clang-tidy/checks/misc/unused-using-decls>` check to avoid false positive when
-  using in elaborated type and only check cpp files.
```
revert this small change.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/move-shared-pointer-contents.rst:
```diff
@@ -0,0 +1,19 @@
+``std::move(*p)`` or ``std::move(*p.get())``. Other reference holders
+may not be expecting the move and suddenly getting empty or otherwise
+indeterminate states can cause issues.
+
```
add info that check is restricted to C++11 and above.

PiotrZSL reviewed (changes requested):
@pizzud
To be honest I see this check being already done in 99%.
Please resolve some last nits, rebase code and we could merge it in this month.

pizzud committed:
Merge branch 'shared-ptr' of https://github.com/pizzud/llvm-project into shared-ptr

pizzud committed:
fix merge conflicts

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,145 @@
+          Options.get("SharedPointerClasses", "::std::shared_ptr"))) {}
+
+void MoveSharedPointerContentsCheck::registerMatchers(MatchFinder *Finder) {
+  auto isStdMove = callee(functionDecl(hasName("::std::move")));
```
Good idea! Done.

PiotrZSL commented:
Looks like there are still some merge conflicts, duplicated lines and so on.

pizzud committed:
merge

pizzud committed:
cleanup duplicate entry

pizzud committed:
cleanup a dupe and restore release notes

pizzud committed:
avoid merge conflict

pizzud committed:
Merge branch 'main' into shared-ptr

pizzud commented:
Should be resolved now. Thanks for your patience.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/move-shared-pointer-contents.rst:
```diff
@@ -0,0 +1,22 @@
+   A semicolon-separated list of class names that should be treated as
+   shared pointers. Classes are resolved through aliases, so any alias
+   to the defined classes will be considered. Default is
+   `::std::shared_ptr` and `::boost::shared_pointer`.
```
```suggestion
   `::std::shared_ptr;::boost::shared_pointer`.
```

pizzud committed:
Address formatting suggestion in checker doc.

Co-authored-by: EugeneZelenko <eugene.zelenko@gmail.com>

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,142 @@
+              0,
+              cxxOperatorCallExpr(
+                  hasOverloadedOperatorName("*"),
+                  hasDescendant(declRefExpr(hasType(qualType(isSharedPointer(
```
this has hasDescendant may not work properly here, as this callExpr would match:
```
std::move(*getSomething(objec1, shared_ptr_obj)
```

Instead you could check type of object used for operator *:

```
cxxOperatorCallExpr(callee(ofClass(matchers::matchesAnyListedName(SharedPointerClasses)))
```

pizzud replied:
Doing that gives me a QualType for the underlying type, not the user-visible type. As an example, this matters when operator* or get() are declared in a parent rather than the visible class, which is the case for at least one major implementation of std::shared_ptr (I think it's libstdc++ but haven't looked in several months) and is visible in the unit tests because the dummy implementation in use has a std::__shared_ptr that defines them and std::shared_ptr inherits from it. Using DeclRefExpr is an intentional attempt to get at the user-visible type of the variable. The hasDescendant can at least be hasArgument(0).

The good news is that I added your example to the unit tests and it doesn't actually trigger.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,142 @@
+              0, unaryOperator(
+                     hasOperatorName("*"),
+                     hasUnaryOperand(allOf(
+                         hasDescendant(declRefExpr(hasType(qualType(
```
same issue here as above

pizzud replied:
Same response, but in this case hasArgument(0) hits a static_assert about right polymorphism.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,142 @@
+void MoveSharedPointerContentsCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  clang::SourceLocation Loc;
+  if (const auto *UnresolvedCall =
+          Result.Nodes.getNodeAs<CallExpr>("unresolved_call");
+      UnresolvedCall != nullptr) {
+    Loc = UnresolvedCall->getBeginLoc();
+  } else if (const auto *UnresolvedGetCall =
+                 Result.Nodes.getNodeAs<CallExpr>("unresolved_get_call");
+             UnresolvedGetCall != nullptr) {
+    Loc = UnresolvedGetCall->getBeginLoc();
+  } else if (const auto *GetCall = Result.Nodes.getNodeAs<CallExpr>("get_call");
+             GetCall != nullptr) {
+    Loc = GetCall->getBeginLoc();
+  } else if (const auto *Call = Result.Nodes.getNodeAs<CallExpr>("call");
+             Call != nullptr) {
+    Loc = Call->getBeginLoc();
+  } else {
+    return;
+  }
```
instead of using different names, just bind those to same name and then you will be able just call getBeginLoc once

PiotrZSL reviewed (changes requested):
Matchers need simplification, mainly that hasDescendant and declRefExpr need to go

pizzud committed:
reflow check() implementation and add another unit test

denzor200 commented:
Suppose this check should be silent for a shared_ptr with `use_count() == 1`. Is it possible to change implementation in order to following this way?

pizzud commented:
> Suppose this check should be silent for a shared_ptr with use_count() == 1. Is it possible to change implementation in order to following this way?

Wouldn't that be a runtime check, and thus not something clang-tidy can do without significant data flow analysis effort?

Separately, I finally noticed the failing unit tests on Windows and will start investigating, though I don't have a Windows box so it might be tricky.

pizzud committed:
Merge branch 'main' into shared-ptr

pizzud committed:
Merge branch 'llvm:main' into shared-ptr

pizzud committed:
fix whitespace and remove test case that mysteriously fails on Windows and provides limited additional coverage

denzor200 commented:
> Wouldn't that be a runtime check, and thus not something clang-tidy can do without significant data flow analysis effort?

I'm not experienced enough in clang-tidy internals, but I believe it will not be a serious challenge to check for having `assert(sp.use_count() == 1)` one line before `*sp` being moved.

I suppose everybody agree that this snippet
```
assert(sp.use_count() == 1);
auto y = std::move(*sp);
```
..is better than that one:
```
// NOLINTNEXTLINE(clang-tidy-bugprone-move-shared-pointer-contents)
auto y = std::move(*sp);
```

vbvictor commented:
@pizzud, you wish to continue work on this PR? 21th release of LLVM would be in late July, we have time to prepare and merge this check before the release. Also, It's okay to not cover all cases in the first iteration, some improvements can be added in later patches.

Windows build is failing most likely to absent of "-fno-delayed-template-parsing" flag in test command, look for it in other tests for reference.

pizzud committed:
Merge branch 'main' into shared-ptr

pizzud committed:
Add flag for Windows fixes

pizzud committed:
Merge branch 'llvm:main' into shared-ptr

vbvictor commented:
Please ping me when you're done working and want to test CI build
I need to approve every CI build/format run... not very convenient.

pizzud committed:
Update clang-tools-extra/docs/clang-tidy/checks/bugprone/move-shared-pointer-contents.rst

Co-authored-by: EugeneZelenko <eugene.zelenko@gmail.com>

pizzud committed:
cleanup includes/whitespace

pizzud committed:
whitespace and formatting fixes

pizzud committed:
whitespace fix

pizzud committed:
whitespace and formatting fixes

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,137 @@
+    }
+  }
+
+  if (Call == nullptr && !Call->getBeginLoc().isValid()) {
```
Should this be `Call != nullptr` or simply `Call &&`

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,137 @@
+
+  if (Call == nullptr && !Call->getBeginLoc().isValid()) {
+    return;
+  }
```
Avoid using braces for single if-stmt

vbvictor commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -148,6 +148,12 @@ New checks
   Finds potentially erroneous calls to ``reset`` method on smart pointers when
   the pointee type also has a ``reset`` method.
 
+- New :doc:`bugprone-move-shared-pointer-contents
```
Please place in alphabetical order by check name

pizzud committed:
Merge branch 'llvm:main' into shared-ptr

pizzud committed:
Merge branch 'llvm:main' into shared-ptr

vbvictor commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/move-shared-pointer-contents.rst:
```diff
@@ -0,0 +1,23 @@
+   A semicolon-separated list of class names that should be treated as
+   shared pointers. Classes are resolved through aliases, so any alias
+   to the defined classes will be considered. Default is
+   `::std::shared_ptr;::boost::shared_pointer`.
```
Boost has `::boost::shared_ptr`, not `::boost::shared_pointer`

pizzud replied:
Oops, done. Fixed in the actual checker as well.

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,136 @@
+#include "../ClangTidyCheck.h"
+#include "../utils/Matchers.h"
+#include "../utils/OptionsUtils.h"
+#include "clang/AST/ASTContext.h"
```
Unused include?
```suggestion
```

pizzud replied:
Done.

vbvictor reviewed (commented):
Could you please run clang-tidy over your file and fix the errors, clangd can detect them, too.

vbvictor commented:
I played with matchers a bit, one `hasDescendant` may be eliminated by:

```diff
diff --git a/clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp b/clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp
index 4d5c25bf9f27..ab5e13fbb9cc 100644
--- a/clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp
+++ b/clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp
@@ -70,12 +70,11 @@ void MoveSharedPointerContentsCheck::registerMatchers(MatchFinder *Finder) {
           hasArgument(
               0, unaryOperator(
                      hasOperatorName("*"),
-                     hasUnaryOperand(allOf(
-                         hasDescendant(declRefExpr(hasType(qualType(
-                             isSharedPointer(matchers::matchesAnyListedName(
-                                 SharedPointerClasses)))))),
-                         cxxMemberCallExpr(
-                             callee(cxxMethodDecl(hasName("get")))))))))
+                     hasUnaryOperand(cxxMemberCallExpr(
+                         callee(cxxMethodDecl(hasName("get"))),
+                         on(hasType(qualType(isSharedPointer(
+                             matchers::matchesAnyListedName(
+                                 SharedPointerClasses))))))))))
           .bind("get_call")));
```

pizzud committed:
Fix lint errors and Boost shared_ptr name

pizzud committed:
Merge branch 'llvm:main' into shared-ptr

pizzud reviewed (commented):
I think I got the clang-tidy issues sorted out.

vbvictor reviewed (commented):
Could you also try to eliminate second `hasDescendant(declRefExpr())` in similar way if possible. I'd try to look at it this week too.
I suppose it's the only issue to be resolved before merge.

pizzud committed:
usings and whitespace

pizzud committed:
Merge branch 'llvm:main' into shared-ptr

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,134 @@
+#include "../utils/OptionsUtils.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers; // NOLINT
```
Unless I'm missing something, the alternative is ~27 individual `using`s for matchers. I don't particularly love it, but it does work and I can at least isolate it to the one function.

vbvictor replied:
Oh, sorry, it should be this one `using` instead of individual ones. I was confused by NOLINT. clang-tidy shouldn't warn on this line. Could you please revert the change

pizzud commented on clang-tools-extra/clang-tidy/bugprone/MoveSharedPointerContentsCheck.cpp:
```diff
@@ -0,0 +1,134 @@
+      callExpr(IsStdMove,
+               hasArgument(
+                   0, unaryOperator(hasOperatorName("*"),
+
```
Done.

pizzud reviewed (commented):
I'll give the matcher simplification a shot and see how it goes; I have a very short working week this week so I'm unlikely to make much progress.

pizzud committed:
fix using since clang-tidy should allow this

pizzud committed:
Merge branch 'llvm:main' into shared-ptr
